postprocessing/science/spectral_corner_frequency_from_surface_xdmf.py

```diff
@@ -0,0 +1,384 @@
+    brune_function = q_zero*(1/(1+(freq_tensor/fc_tensor)**n))
+    
+    # Calculate the misfit of every (Q, fc) combination
+    if misfit == "RMS":
+        misfit_tensor = (brune_function - amplitude_tensor)**2
+        misfit_tensor = np.sqrt(np.mean(misfit_tensor, axis=0))
+    if misfit == "SSM":
+        misfit_tensor = np.abs(np.log10(amplitude_tensor/brune_function))
+        misfit_tensor = np.mean(misfit_tensor, axis=0)
+    
+    # Get indices of minimum misfit (Q, fc) combination
+    index = np.unravel_index(np.argmin(misfit_tensor), misfit_tensor.shape)
+    
+    fc = fc_tensor[0,index[0],0]
+    if q_dim != 0:
+        q = q_zero[0,0,index[1]]
+    else:
+        q = q_zero[0,index[1],0]
+        
+    return fc, q
+
+def complete_fft(tr, dt):
+    tr_fft=np.fft.rfft(tr.data)
+    times_comp = np.fft.rfftfreq(tr.shape[0],dt)
+    return tr_fft, times_comp
+
+def ApproximateEventDurationAndHypocenter(slipRateThreshold=args.slipRateThreshold[0]):
+    """Calculates the derivative of ASl (probably faster than loading SR1 and SR2)
+     and approximates event duration by multiplying dt with the number of timesteps,
+     where maximum on-fault slip rate is above a threshold (slipRateThreshold)"""
+    
+    if nprocs == 1 or not args.parallelLoading:
+        ASl = faultxdmf.ReadData("ASl").T[::stepsize,timeIndicesFault[0]:timeIndicesFault[1]]
+    else:
+        ASl = LoadSingleComponentParallel("ASl", surfaceData=False)
+    ASR = np.gradient(ASl, dtFault, axis=1)
+    duration = np.sum(np.where(np.amax(ASR, axis=0) > slipRateThreshold, 1, 0))*dtFault
+    slippingFault = np.where(ASR > slipRateThreshold, 1, 0)
+    indHypocenter = np.argmax(slippingFault[:,np.argmax(np.amax(slippingFault, axis=0))])
+    hypocenter = faultxyz[indHypocenter,:]
+    return duration, hypocenter
+
+def PickPPhases(trigger=0.001):
+    """Picks the P-wave arrival time for each receiver. This function is only suited for synthetic data without noise"""
+    
+    data = np.sqrt(u**2 + v**2 + w**2)
+    thresholds = np.amax(data, axis=1)*trigger  # trigger dictates the portion of the maximum ground velocity,
+    thresholds = thresholds.reshape(-1,1)       # which needs to be reached to pick the p-phase
+    ind = np.argmax(data>thresholds, axis=1)
+    t = ind*dt
+    return np.transpose(np.vstack((ind,t))) # p-phase output: first column index and second column simulation time
+
+def SWaveDelay(xyz, pphase, hypocenter, centroid):
+    """Approximates the difference between P- and S-wave arrivals for each receiver"""
+    
+    if not args.avgSWaveVelocity:
+        distance1 = np.linalg.norm((xyz - hypocenter), axis=1)
+        swaveVelocity = np.mean(distance1 / pphase) / np.sqrt(3)
+    else:
+        swaveVelocity = args.avgSWaveVelocity[0]
+    print("Average S-wave velocity near the surface: "+str(swaveVelocity))
+    distance2 = np.linalg.norm((xyz - centroid), axis=1)
+    swaveDelay = distance2 / (swaveVelocity * (np.sqrt(3) - 1))
```

Comment:
d/Vs - d/Vp = d/Vs - d/(sqrt(3)Vs)    = d/Vs * (1 - 1/sqrt(3)) = 0.42264973081037416 d/Vs
which is not the same as 1/(sqrt(3)-1)  d/Vs =1.3660254037844388  d/Vs ?

Reply:
Ups, I corrected it